tools/clang/include/clang/SPIRV/SpirvModule.h

```diff
@@ -168,11 +205,53 @@ class SpirvModule {
   std::vector<SpirvVariable *> variables;
   std::vector<SpirvFunction *> functions;
 
-  // DebugScope, DebugNoScope, DebugDeclare, DebugValue instructions can
-  // interleave with instructions within a function body. All other debugging
-  // instructions should be located between section 9 and section 10.
-  // This vector contains all of these instructions which fall between section 9
-  // and section 10.
+  // We should emit debug info instructions in the order of
+  // 1. DebugOperation
+  // 2. DebugExpression
+  // 3. For each DebugSource
+  // 4. DebugCompilationUnit
+  // 5. DebugTypeXXX
+  // 6. DebugGlobalVariable
+  // 7. For each DebugFunction
+  //    7.1. DebugTypeXXX defined in this function
+  //    7.2. DebugGlobalVariable or DebugLocalVariable defined in
+  //         this function
+  //    7.3. For each DebugLexicalBlock defined in this function
+  //         7.3.1. DebugTypeXXX defined in block
+  //         7.3.2. DebugLocalVariable defined in block
+  // 8. DebugTypeMember and DebugTypeInheritance
+  //
+  // If we do not follow this order, generated SPIR-V code would
+  // be invalid because of forward references.
```

Comment:
I think we should fix this with the concept of placeholders.  While the final SPIRV cannot have forward references, we should be able to build the web of debug info with placeholders, which are then solved before final SPIRV is emitted.  For now, we can leave this as-is.  But relying on a fixed generation order is generally a problem long term.

Reply:
I added a TODO here.